### ISSUE TITLE
ImageResize and AvatarOptimize fail

### ISSUE THREAD

**Issue by matrix07012** [pixelfed/support#55]
I'm using ImageMagick and both ImageResize and AvatarOptimize fail for me because they can't write image data to path.

![image](https://user-images.githubusercontent.com/8049717/66203292-47805e00-e6a8-11e9-8b51-c902ae20ff69.png)
![image](https://user-images.githubusercontent.com/8049717/66203322-61ba3c00-e6a8-11e9-83fd-105513245f56.png)
![image](https://user-images.githubusercontent.com/8049717/66203348-71d21b80-e6a8-11e9-8558-0bb18229399b.png)

**Comment by trwnh** [comment]
OK, you need to make the group have write permissions -- 664 on files, not 644. And 775 on directories, not 755. https://docs.pixelfed.org/installing-pixelfed/installation.html#set-correct-permissions

**Comment by matrix07012** [comment]
@trwnh That's the issue. I did this at the start and all the other folders and files have correct permissions.
Now I tried doing this again and that fixes it only for already existing files and folders, but if I upload a new picture, that picture has wrong permissions or make a new user, their folder has wrong permissions.

**Comment by trwnh** [comment]
Hmm. I'd like to test two things. Can you run the following and test an upload after?

- setgid: `chmod -R g+s /path/to/pixelfed`
- umask: `echo "umask 002" > /home/pixelfed/.profile`

And of course, to reset your perms if necessary:
```
cd /path/to/pixelfed
chown -R pixelfed:pixelfed .
find . -type d -exec chmod 775 {} \; 
find . -type f -exec chmod 664 {} \; 
```

**Comment by matrix07012** [comment]
setgid and umask 002 only changed the group of the newly uploaded file, owner is still web-data, thus the job fails. Wouldn't setuid make more sense?
I think that the second command is supposed to be `echo "umask 002" >> /home/pixelfed/.profile`

**Comment by trwnh** [comment]
The end goal is to have directories and files automatically created as 4775 and 4664 by pixelfed:pixelfed ideally. I'm not sure what the difference between setuid and setgid does, practically speaking. The issue is that there are plenty of valid ways it could be done, so there's no one right answer. Consequently, there are also a lot of ways it could be done wrong.

For the app-user, at least, this can be done either by setting their umask to 002 (if running horizon directly from a shell), or by using the `User`, `Group`, and `UMask` directives within the systemd service (if using systemd service to automatically run horizon in the background).

**Comment by matrix07012** [comment]
I tried setuid and it did nothing, but I don't think that Horizon plays any role in uploading since you can upload pictures without it running.

**Comment by trwnh** [comment]
Yeah, I think the umask thing is more important here, since you've already got the group membership set up both ways... you just need the default permissions to be umasked to 002 instead of 022, so that the group members have write perms.

I just don't remember having to do any sort of umask for php-fpm when setting up my dev env, and everything works... I did set up a umask for my pixelfed app-user, though, and I'm trying to figure out if editing the systemd service might be better than messing wiith .profile

**Comment by matrix07012** [comment]
Setting UMask=0002 in the systemd service didn't help. Were you using Apache or Nginx when you were testing this? I'm using Nginx.

**Comment by trwnh** [comment]
I was using nginx too, on arch. Using the http user as the app user is the simpler "default" option if you read the docs (and I did it that way because I'm not 100% confident in how to get different users working for everyone yet), but I'm still not entirely sure why I got multiple users working and you didn't.

Anyway, I did some playing around in a shell to try and figure out how g+s and umask interact, and here's what I've got:

- I think you've got the group ownership being maintained correctly by setgid, but...
- One of your processes is not setting group write perms correctly, as it should with umask.

If you have your pixelfed user set up to allow shell logins, you can try logging into a shell and using something like `touch testfile` to verify if you have the .profile set up correctly. Or you can set it up systemwide by editing `/etc/profile` directly. I think it's possible that the issue is with the nginx systemd service needing a UMask directive? But I don't remember doing so myself.

**Comment by matrix07012** [comment]
I think I fixed it by running php fmp under pixeled user , but I'm not sure if it's a good solution.

Btw if you are going to change documentation, the Nginx config is missing `client_max_body_size`.

**Comment by trwnh** [comment]
Idk if setting php-fpm user to pixelfed is the best solution -- if you're fine with using that machine only for pixelfed, then the better solution is to edit /etc/profile and set the umask for the whole system.

**Comment by trwnh** [comment]
I think the best solution for a machine with multiple sites is to actually create a separate PHP-FPM pool: https://wiki.archlinux.org/index.php/User:Fordprefect/php-fpm#Using_pools_for_site_isolation

I still need to make sure I have the write config, but I think I'll update the docs to set up `/etc/php/php-fpm.d/pixelfed.conf`:

```
#     use the username of the app-user as the pool name, e.g. pixelfed
[pixelfed]
listen = /run/php-fpm/pixelfed.sock;
#    to use a tcp socket instead, e.g. if running php-fpm on a different machine than your app:
#    (note that the port 9001 is used, since php-fpm defaults to running on port 9000;)
#    (however, the port can be whatever you want)
# listen = 127.0.0.1:9001;
user = pixelfed
group = pixelfed
listen.owner = http
listen.group = http
listen.mode = 0660
```

Then in the nginx conf:

```
fastcgi_pass unix:/run/php-fpm/pixelfed.sock;
```
